// @mux/mux-node/resources/video/transcription-vocabularies.js
"use strict";
// File generated from our OpenAPI spec by Stainless.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TranscriptionVocabulariesBasePage = exports.TranscriptionVocabularies = void 0;
const resource_1 = require("@mux/mux-node/resource");
const core_1 = require("@mux/mux-node/core");
const TranscriptionVocabulariesAPI = __importStar(require("@mux/mux-node/resources/video/transcription-vocabularies"));
const pagination_1 = require("@mux/mux-node/pagination");
class TranscriptionVocabularies extends resource_1.APIResource {
    /**
     * Create a new Transcription Vocabulary.
     */
    create(body, options) {
        return this._client.post('/video/v1/transcription-vocabularies', { body, ...options })._thenUnwrap((obj) => obj.data);
    }
    /**
     * Retrieves the details of a Transcription Vocabulary that has previously been
     * created. Supply the unique Transcription Vocabulary ID and Mux will return the
     * corresponding Transcription Vocabulary information. The same information is
     * returned when creating a Transcription Vocabulary.
     */
    retrieve(transcriptionVocabularyId, options) {
        return this._client.get(`/video/v1/transcription-vocabularies/${transcriptionVocabularyId}`, options)._thenUnwrap((obj) => obj.data);
    }
    /**
     * Updates the details of a previously-created Transcription Vocabulary. Updates to
     * Transcription Vocabularies are allowed while associated live streams are active.
     * However, updates will not be applied to those streams while they are active.
     */
    update(transcriptionVocabularyId, body, options) {
        return this._client.put(`/video/v1/transcription-vocabularies/${transcriptionVocabularyId}`, {
            body,
            ...options,
        })._thenUnwrap((obj) => obj.data);
    }
    list(query = {}, options) {
        if ((0, core_1.isRequestOptions)(query)) {
            return this.list({}, query);
        }
        return this._client.getAPIList('/video/v1/transcription-vocabularies', TranscriptionVocabulariesBasePage, { query, ...options });
    }
    /**
     * Deletes a Transcription Vocabulary. The Transcription Vocabulary's ID will be
     * disassociated from any live streams using it. Transcription Vocabularies can be
     * deleted while associated live streams are active. However, the words and phrases
     * in the deleted Transcription Vocabulary will remain attached to those streams
     * while they are active.
     */
    delete(transcriptionVocabularyId, options) {
        return this._client.delete(`/video/v1/transcription-vocabularies/${transcriptionVocabularyId}`, {
            ...options,
            headers: { Accept: '*/*', ...options?.headers },
        });
    }
}
exports.TranscriptionVocabularies = TranscriptionVocabularies;
class TranscriptionVocabulariesBasePage extends pagination_1.BasePage {
}
exports.TranscriptionVocabulariesBasePage = TranscriptionVocabulariesBasePage;
(function (TranscriptionVocabularies) {
    TranscriptionVocabularies.TranscriptionVocabulariesBasePage = TranscriptionVocabulariesAPI.TranscriptionVocabulariesBasePage;
})(TranscriptionVocabularies = exports.TranscriptionVocabularies || (exports.TranscriptionVocabularies = {}));
