// @mux/mux-node/resources/video/uploads.js
"use strict";
// File generated from our OpenAPI spec by Stainless.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UploadsBasePage = exports.Uploads = void 0;
const resource_1 = require("@mux/mux-node/resource");
const core_1 = require("@mux/mux-node/core");
const UploadsAPI = __importStar(require("@mux/mux-node/resources/video/uploads"));
const pagination_1 = require("@mux/mux-node/pagination");
class Uploads extends resource_1.APIResource {
    /**
     * Creates a new direct upload, through which video content can be uploaded for
     * ingest to Mux.
     */
    create(body, options) {
        return this._client.post('/video/v1/uploads', { body, ...options })._thenUnwrap((obj) => obj.data);
    }
    /**
     * Fetches information about a single direct upload in the current environment.
     */
    retrieve(uploadId, options) {
        return this._client.get(`/video/v1/uploads/${uploadId}`, options)._thenUnwrap((obj) => obj.data);
    }
    list(query = {}, options) {
        if ((0, core_1.isRequestOptions)(query)) {
            return this.list({}, query);
        }
        return this._client.getAPIList('/video/v1/uploads', UploadsBasePage, { query, ...options });
    }
    /**
     * Cancels a direct upload and marks it as cancelled. If a pending upload finishes
     * after this request, no asset will be created. This request will only succeed if
     * the upload is still in the `waiting` state.
     */
    cancel(uploadId, options) {
        return this._client.put(`/video/v1/uploads/${uploadId}/cancel`, options)._thenUnwrap((obj) => obj.data);
    }
}
exports.Uploads = Uploads;
class UploadsBasePage extends pagination_1.BasePage {
}
exports.UploadsBasePage = UploadsBasePage;
(function (Uploads) {
    Uploads.UploadsBasePage = UploadsAPI.UploadsBasePage;
})(Uploads = exports.Uploads || (exports.Uploads = {}));
