// @mux/mux-node/resources/video/video.mjs
// File generated from our OpenAPI spec by Stainless.
import { APIResource } from '@mux/mux-node/resource';
import * as AssetsAPI from '@mux/mux-node/resources/video/assets';
import * as DeliveryUsageAPI from '@mux/mux-node/resources/video/delivery-usage';
import * as LiveStreamsAPI from '@mux/mux-node/resources/video/live-streams';
import * as PlaybackIDsAPI from '@mux/mux-node/resources/video/playback-ids';
import * as PlaybackRestrictionsAPI from '@mux/mux-node/resources/video/playback-restrictions';
import * as SpacesAPI from '@mux/mux-node/resources/video/spaces';
import * as TranscriptionVocabulariesAPI from '@mux/mux-node/resources/video/transcription-vocabularies';
import * as UploadsAPI from '@mux/mux-node/resources/video/uploads';
export class Video extends APIResource {
    constructor() {
        super(...arguments);
        this.assets = new AssetsAPI.Assets(this._client);
        this.deliveryUsage = new DeliveryUsageAPI.DeliveryUsage(this._client);
        this.liveStreams = new LiveStreamsAPI.LiveStreams(this._client);
        this.playbackIds = new PlaybackIDsAPI.PlaybackIDs(this._client);
        this.playbackRestrictions = new PlaybackRestrictionsAPI.PlaybackRestrictions(this._client);
        this.spaces = new SpacesAPI.Spaces(this._client);
        this.transcriptionVocabularies = new TranscriptionVocabulariesAPI.TranscriptionVocabularies(this._client);
        this.uploads = new UploadsAPI.Uploads(this._client);
    }
}
(function (Video) {
    Video.Assets = AssetsAPI.Assets;
    Video.AssetsBasePage = AssetsAPI.AssetsBasePage;
    Video.DeliveryUsage = DeliveryUsageAPI.DeliveryUsage;
    Video.DeliveryReportsPageWithTotal = DeliveryUsageAPI.DeliveryReportsPageWithTotal;
    Video.LiveStreams = LiveStreamsAPI.LiveStreams;
    Video.LiveStreamsBasePage = LiveStreamsAPI.LiveStreamsBasePage;
    Video.PlaybackIDs = PlaybackIDsAPI.PlaybackIDs;
    Video.PlaybackRestrictions = PlaybackRestrictionsAPI.PlaybackRestrictions;
    Video.PlaybackRestrictionsBasePage = PlaybackRestrictionsAPI.PlaybackRestrictionsBasePage;
    Video.Spaces = SpacesAPI.Spaces;
    Video.SpacesBasePage = SpacesAPI.SpacesBasePage;
    Video.TranscriptionVocabularies = TranscriptionVocabulariesAPI.TranscriptionVocabularies;
    Video.TranscriptionVocabulariesBasePage = TranscriptionVocabulariesAPI.TranscriptionVocabulariesBasePage;
    Video.Uploads = UploadsAPI.Uploads;
    Video.UploadsBasePage = UploadsAPI.UploadsBasePage;
})(Video || (Video = {}));
