// @mux/mux-node/resources/system/signing-keys.mjs
// File generated from our OpenAPI spec by Stainless.
import { APIResource } from '@mux/mux-node/resource';
import { isRequestOptions } from '@mux/mux-node/core';
import * as SigningKeysAPI from '@mux/mux-node/resources/system/signing-keys';
import { BasePage } from '@mux/mux-node/pagination';
export class SigningKeys extends APIResource {
    /**
     * Creates a new signing key pair. When creating a new signing key, the API will
     * generate a 2048-bit RSA key-pair and return the private key and a generated
     * key-id; the public key will be stored at Mux to validate signed tokens.
     */
    create(options) {
        return this._client.post('/system/v1/signing-keys', options)._thenUnwrap((obj) => obj.data);
    }
    /**
     * Retrieves the details of a signing key that has previously been created. Supply
     * the unique signing key ID that was returned from your previous request, and Mux
     * will return the corresponding signing key information. **The private key is not
     * returned in this response.**
     */
    retrieve(signingKeyId, options) {
        return this._client.get(`/system/v1/signing-keys/${signingKeyId}`, options)._thenUnwrap((obj) => obj.data);
    }
    list(query = {}, options) {
        if (isRequestOptions(query)) {
            return this.list({}, query);
        }
        return this._client.getAPIList('/system/v1/signing-keys', SigningKeysBasePage, { query, ...options });
    }
    /**
     * Deletes an existing signing key. Use with caution, as this will invalidate any
     * existing signatures and no JWTs can be signed using the key again.
     */
    delete(signingKeyId, options) {
        return this._client.delete(`/system/v1/signing-keys/${signingKeyId}`, {
            ...options,
            headers: { Accept: '*/*', ...options?.headers },
        });
    }
}
export class SigningKeysBasePage extends BasePage {
}
(function (SigningKeys) {
    SigningKeys.SigningKeysBasePage = SigningKeysAPI.SigningKeysBasePage;
})(SigningKeys || (SigningKeys = {}));
